### PR TITLE
fix(glm): apply LoRA to TileLang MLA KV projection

### PR DESCRIPTION
**Pull request by shahafwa**
# What does this PR do?

Fixes missing LoRA application in the GLM DSA TileLang attention path.

The TileLang path directly accessed and reshaped `self.kv_b_proj.weight`. When `kv_b_proj` was wrapped by `LinearLoRA`, this bypassed the LoRA effective-weight calculation:

`W_effective = W_base + scale * (B @ A)`

Consequently, the KV LoRA update did not affect TileLang attention outputs, and the adapter parameters did not receive the expected gradients.

This change adds differentiable effective-weight materialization for ordinary Torch LoRA layers and uses that effective weight in the TileLang absorbed KV calculation.
Unsupported cases such as active training dropout, DoRA, and quantized or delegated linear implementations fail explicitly instead of silently producing incorrect behavior.

# Changelog

- Add differentiable LoRA effective-weight materialization.
- Use the effective `kv_b_proj` weight in GLM TileLang MLA.
- Add forward-output and backward-gradient parity coverage.
- Verify that both absorbed K and projected V use the LoRA weight.
- Add a cold-start optimizer test proving that zero-initialized LoRA B updates first and subsequently enables gradients for LoRA A.

# Testing

- LoRA unit tests: 26 passed, 4 optional skips.
- GLM TileLang tests: 33 passed, 6 optional GPU/kernel skips.
- Ruff formatting and lint checks passed.

# Additional Information

This fixes the TileLang-specific path that read the raw base weight. The normal dense attention path already invokes `kv_b_proj` as a module and therefore already applies LoRA.

### CODE REVIEW

**Comment by HuiyingLi** [comment]
Thank you so much @shahafwa  for the fix! I appreciate it! Do you mind also attaching a glm5.2 lora config yaml? Just realize we don't have it yet :D

**Comment by shahafwa** [comment]
@HuiyingLi sure 
```
  recipe: TrainFinetuneRecipeForNextTokenPrediction

  seed: 42
  longalign: false

  step_scheduler:
    global_batch_size: 32
    local_batch_size: 1
    ckpt_every_steps: 25
    val_every_steps: 5
    gc_every_steps: 25
    num_epochs: 1
    save_checkpoint_every_epoch: true

  distributed:
    strategy: fsdp2
    tp_size: 1
    cp_size: 8
    pp_size: 1
    ep_size: 8
    sequence_parallel: false
    activation_checkpointing: true
    reshard_after_forward: false
    moe:
      reshard_after_forward: true
      wrap_outer_model: false
      ignore_router_for_ac: true

  dist_env:
    backend: nccl
    timeout_minutes: 240

  model:
    _target_: nemo_automodel.NeMoAutoModelForCausalLM.from_pretrained
    pretrained_model_name_or_path: ${oc.env:GLM52_MODEL_PATH}
    trust_remote_code: true
    config:
      head_dim: 64
    backend:
      _target_: nemo_automodel.components.models.common.BackendConfig
      attn: tilelang
      linear: torch
      rms_norm: torch_fp32
      rope_fusion: false
      experts: torch
      experts_storage: fp8
      dispatcher: torch
      gate_precision: float32
      fake_balanced_gate: false
      enable_hf_state_dict_adapter: true
      enable_fsdp_optimizations: true

  peft:
    _target_: nemo_automodel.components._peft.lora.PeftConfig
    target_modules:
      - "*.self_attn.q_a_proj"
      - "*.self_attn.q_b_proj"
      - "*.self_attn.kv_a_proj_with_mqa"
      - "*.self_attn.kv_b_proj"
      - "*.self_attn.o_proj"
      - "*.mlp.gate_proj"
      - "*.mlp.up_proj"
      - "*.mlp.down_proj"
      - "*.mlp.shared_experts.gate_proj"
      - "*.mlp.shared_experts.up_proj"
      - "*.mlp.shared_experts.down_proj"
      - "*.mlp.experts"
    dim: 32
    alpha: 64
    dropout: 0.0
    use_memory_efficient_lora: true
    use_triton: false

  checkpoint:
    enabled: true
    checkpoint_dir: ${oc.env:GLM52_CHECKPOINT_DIR}
    restore_from: ${oc.env:GLM52_RESTORE_FROM,}
    save_consolidated: false
    dequantize_base_checkpoint: true

  clip_grad_norm:
    max_norm: 1.0

  loss_fn:
    _target_: nemo_automodel.components.loss.chunked_ce.ChunkedCrossEntropy
    chunk_len: 512
    compile: true

  dataset:
    _target_: nemo_automodel.components.datasets.llm.chat_dataset.ChatDataset
    path_or_dataset_id: ${oc.env:GLM52_DATASET_PATH}
    split: train
    parse_tool_call_arguments: true
    reasoning_effort_column: reasoning_effort
    clear_thinking: false
    truncation: false
    seq_length: 200704
    padding: false

  validation_dataset:
    _target_: nemo_automodel.components.datasets.llm.chat_dataset.ChatDataset
    path_or_dataset_id: ${oc.env:GLM52_VALIDATION_DATASET_PATH}
    split: train
    parse_tool_call_arguments: true
    reasoning_effort_column: reasoning_effort
    clear_thinking: false
    truncation: false
    seq_length: 200704
    padding: false

  packed_sequence:
    packed_sequence_size: 0

  dataloader:
    _target_: torchdata.stateful_dataloader.StatefulDataLoader
    collate_fn:
      _target_: nemo_automodel.components.datasets.utils.packed_sequence_thd_collater
      pad_seq_len_divisible: 4096
    shuffle: true
    drop_last: false
    num_workers: 0
    pin_memory: false

  validation_dataloader:
    _target_: torchdata.stateful_dataloader.StatefulDataLoader
    collate_fn:
      _target_: nemo_automodel.components.datasets.utils.packed_sequence_thd_collater
      pad_seq_len_divisible: 4096
    shuffle: false
    drop_last: false
    num_workers: 0
    pin_memory: false

  optimizer:
    _target_: torch.optim.AdamW
    betas:
      - 0.9
      - 0.95
    eps: 1.0e-8
    lr: 1.0e-4
    weight_decay: 0

  lr_scheduler:
    lr_warmup_steps: 0
    lr_decay_style: constant
    min_lr: 2.0e-5
```

**Comment by HuiyingLi** [comment]
@shahafwa Thank you so much for sharing! Is it possible to commit the yaml into this PR? Thank you!

**Comment by shahafwa** [comment]
done  :D @HuiyingLi

**Comment by HuiyingLi** [comment]
Thank you so much! Would you mind fixing the DCO and I will trigger tests

File: examples/llm_finetune/glm/glm_5.2_lora.yaml [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    wrap_outer_model: false
+    ignore_router_for_ac: true
+
+dist_env:
+  backend: nccl
+  timeout_minutes: 240
+
+model:
+  _target_: nemo_automodel.NeMoAutoModelForCausalLM.from_pretrained
+  pretrained_model_name_or_path: ${oc.env:GLM52_MODEL_PATH}
```
Comment (HuiyingLi) [inline examples/llm_finetune/glm/glm_5.2_lora.yaml]: We could use the glm5.2 model ID here

Comment (shahafwa) [reply examples/llm_finetune/glm/glm_5.2_lora.yaml]: on it thanks!

File: examples/llm_finetune/glm/glm_5.2_lora.yaml [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  max_norm: 1.0
+
+loss_fn:
+  _target_: nemo_automodel.components.loss.chunked_ce.ChunkedCrossEntropy
+  chunk_len: 512
+  compile: true
+
+dataset:
+  _target_: nemo_automodel.components.datasets.llm.chat_dataset.ChatDataset
+  path_or_dataset_id: ${oc.env:GLM52_DATASET_PATH}
```
Comment (HuiyingLi) [inline examples/llm_finetune/glm/glm_5.2_lora.yaml]: We'd need a real dataset ID so that our ci can directly run and test this config. Would tulu3 dataset work here?

File: examples/llm_finetune/glm/glm_5.2_lora.yaml [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  parse_tool_call_arguments: true
+  reasoning_effort_column: reasoning_effort
+  clear_thinking: false
+  truncation: false
+  seq_length: 200704
+  padding: false
+
+validation_dataset:
+  _target_: nemo_automodel.components.datasets.llm.chat_dataset.ChatDataset
+  path_or_dataset_id: ${oc.env:GLM52_VALIDATION_DATASET_PATH}
```
Comment (HuiyingLi) [inline examples/llm_finetune/glm/glm_5.2_lora.yaml]: Same as the above reason :D

File: examples/llm_finetune/glm/glm_5.2_lora.yaml (hunk 2 above)
Comment (shahafwa) [reply examples/llm_finetune/glm/glm_5.2_lora.yaml]: we can try for sure :)

**Comment by shahafwa** [comment]
@HuiyingLi all changes are in.

File: examples/llm_finetune/glm/glm_5.2_lora.yaml [hunk 4]
```diff
@@ -0,0 +1,163 @@
+# LoRA fine-tuning recipe for GLM-5.2 using the TileLang MLA attention backend.
+# Single node: tp=1, cp=8, pp=1, ep=8 -> ep/cp = 1 data-parallel replica => 8 GPUs (1 node).
```
Comment (HuiyingLi) [inline examples/llm_finetune/glm/glm_5.2_lora.yaml]: Match the full-model topology we validated.

```suggestion
# 16 nodes: tp=1, cp=1, pp=1, ep=128 => 128 GPUs (8 H100/node).
```

File: examples/llm_finetune/glm/glm_5.2_lora.yaml [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  gc_every_steps: 25
+  num_epochs: 1
+  save_checkpoint_every_epoch: true
+
+distributed:
+  strategy: fsdp2
+  tp_size: 1
+  cp_size: 8
+  pp_size: 1
+  ep_size: 8
```
Comment (HuiyingLi) [inline examples/llm_finetune/glm/glm_5.2_lora.yaml]: Use CP=1 and EP=128 for the full-model LoRA run.

```suggestion
  cp_size: 1
  pp_size: 1
  ep_size: 128
```

File: examples/llm_finetune/glm/glm_5.2_lora.yaml [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    wrap_outer_model: false
+    ignore_router_for_ac: true
+
+dist_env:
+  backend: nccl
+  timeout_minutes: 240
+
+model:
+  _target_: nemo_automodel.NeMoAutoModelForCausalLM.from_pretrained
+  pretrained_model_name_or_path: zai-org/GLM-5.2
```
Comment (HuiyingLi) [inline examples/llm_finetune/glm/glm_5.2_lora.yaml]: Keep this as a full model load. Locally, set `GLM52_MODEL_PATH` to the full `glm52_fixed_config`; this recipe should not add a `num_hidden_layers` override.

```suggestion
  pretrained_model_name_or_path: ${oc.env:GLM52_MODEL_PATH,zai-org/GLM-5.2}
```

File: examples/llm_finetune/glm/glm_5.2_lora.yaml [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    head_dim: 64
+  backend:
+    _target_: nemo_automodel.components.models.common.BackendConfig
+    attn: tilelang
+    linear: torch
+    rms_norm: torch_fp32
+    rope_fusion: false
+    experts: torch
+    experts_storage: fp8
+    dispatcher: torch
```
Comment (HuiyingLi) [inline examples/llm_finetune/glm/glm_5.2_lora.yaml]: Use the MoE backend pair from the TileLang full-model run.

```suggestion
    experts: torch_mm
    experts_storage: fp8
    dispatcher: hybridep
```

File: examples/llm_finetune/glm/glm_5.2_lora.yaml [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  dropout: 0.0
+  use_memory_efficient_lora: true
+  use_triton: false
+
+checkpoint:
+  enabled: true
+  checkpoint_dir: ${oc.env:GLM52_CHECKPOINT_DIR,checkpoints}
+  restore_from: ${oc.env:GLM52_RESTORE_FROM,}
+  save_consolidated: false
+  dequantize_base_checkpoint: true
```
Comment (HuiyingLi) [inline examples/llm_finetune/glm/glm_5.2_lora.yaml:85]: For this LoRA recipe, checkpointing can stay off by default.

```suggestion
checkpoint:
  enabled: false
  dequantize_base_checkpoint: true
```

File: examples/llm_finetune/glm/glm_5.2_lora.yaml [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    - 0.9
+    - 0.95
+  eps: 1.0e-8
+  lr: 1.0e-4
+  weight_decay: 0
+
+lr_scheduler:
+  lr_warmup_steps: 0
+  lr_decay_style: constant
+  min_lr: 2.0e-5
```
Comment (HuiyingLi) [inline examples/llm_finetune/glm/glm_5.2_lora.yaml:154]: Add W&B placeholders without baking in a local run name.

```suggestion
lr_scheduler:
  lr_warmup_steps: 0
  lr_decay_style: constant
  min_lr: 2.0e-5

# wandb:
#   project: <your_wandb_project>
#   entity: <your_wandb_entity>
#   name: <your_wandb_exp_name>
#   dir: <your_wandb_save_dir>
```

File: examples/llm_finetune/glm/glm_5.2_lora.yaml [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+ci:
+  # cp_size(8) with ep_size(8) => 1 data-parallel replica => 8 GPUs => 1 node (8 H100/node).
+  recipe_owner: shahafwa
+  nodes: 1
+  time: "00:20:00"
```
Comment (HuiyingLi) [inline examples/llm_finetune/glm/glm_5.2_lora.yaml:166]: Keep the recipe metadata consistent with `ep_size: 128`.

```suggestion
ci:
  # cp_size(1) with ep_size(128) => 128 GPUs => 16 nodes (8 H100/node).
  recipe_owner: shahafwa
  nodes: 16
  time: "00:20:00"
```

**Review by HuiyingLi — COMMENTED** [review]
Suggested only the GLM-5.2 LoRA recipe config knobs from the full-model run: cp=1/ep=128, full-model path placeholder without a layer-count override, torch_mm+hybridep, checkpoint off, W&B placeholders, and the matching 16-node metadata.

**Comment by shahafwa** [comment]
@HuiyingLi done

**Comment by HuiyingLi** [comment]
Full GLM-5.2 LoRA training run completed for the full model config (`cp_size=1`, `ep_size=128`, 16 nodes, 100 steps).

W&B: https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/jvbvwbya

Slurm job `14270120` completed with `ExitCode=0:0` in `00:33:38`.

Loss summary by 10-step bucket, averaged over rank0 logged train loss:

| Steps | Avg loss | Last logged loss in bucket |
|---|---:|---:|
| 0-9 | 0.6943 | 0.6083 |
| 10-19 | 0.5857 | 0.5685 |
| 20-29 | 0.5492 | 0.4809 |
| 30-39 | 0.5423 | 0.5062 |
| 40-49 | 0.5593 | 0.5614 |
| 50-59 | 0.5520 | 0.5522 |
| 60-69 | 0.5642 | 0.5498 |
| 70-79 | 0.5432 | 0.5372 |
| 80-89 | 0.5446 | 0.5935 |
| 90-99 | 0.5473 | 0.5515 |

Final logged step: `step 99`, loss `0.5515`.

**Review by HuiyingLi — APPROVED** [review]
Thank you so much for the fix!
